src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs

```diff
@@ -155,22 +158,25 @@ public override void DocumentAdded(HostProject hostProject, HostDocument documen
                 throw new ArgumentNullException(nameof(document));
             }
 
-            _foregroundDispatcher.AssertForegroundThread();
+            _foregroundDispatcher.AssertSpecializedForegroundOrUIThread();
 
-            if (_projects.TryGetValue(hostProject.FilePath, out var entry))
+            lock (_lock)
```

Comment:
All the code in this file is the same. However, in order to account for the possibility that the methods can be either called from the specialized foreground thread or the UI thread, I put any code that writes to a field variable into a lock to prevent concurrent modification.